### DISCUSSION TITLE
[Immunization] V2 Implementation details

### DISCUSSION THREAD

**Discussion by dubdabasoduba** [opensrp/fhircore#2111] [Ideas]
# Immunizations Context

- Health workflows incorporate immunizations for children and sometimes adults.
- V2 of immunizations on FHIR core aims to make some changes to the Task Scheduling and Profile configuration pieces.
- This discussion will define the changes we need to make and explain why they are needed.

# How are immunizations intended to work

- Tasks will be used to represent Immunizations on the FHIR Core. These will be displayed on the patient profile and
  the upcoming encounters screen.
    - The upcoming encounter screen is a new screen that will be accessed via the profile menus.
- The profile screen will only display **DUE** & **OVERDUE** tasks.
    - This means we will have to update the configs to filter all the tasks we fetch for this page.
- All the **UPCOMING** tasks will be displayed on the upcoming encounter screen.
    - The tasks displayed here will be grouped during display.
- The 2 screens below show the **profile** and the **upcoming encounters**.

| **Patient Profile**                                                                                                          | **Upcoming Encounters**                                                                                                      |
|------------------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------|
| <img src="https://user-images.githubusercontent.com/4817531/222598655-295d6a8d-69e9-47a0-9976-1c906a09720f.png" width="400"> | <img src="https://user-images.githubusercontent.com/4817531/222687440-039ee696-1b6f-4c2d-998a-97082871710b.png" width="400"> |

- The upcoming tasks that are 7 days to their **DUE** date will be displayed on the user profile screen.
    - All the upcoming tasks displayed on the profile screen will not be displayed on the encounters screen
- Immunizations are sometimes grouped based on the age they are to start. This means we will have to create different
  cards on both the encounter and profile screens.
- These cards will have a title _(on the left side of the screen)_ and a button _(on the right side of the screen)_.
  They will launch a form that allows the users to record the immunization data for all the **DUE** and **OVERDUE**
  immunizations in the group.
    - Clicking on the individual immunization tasks will also launch a form that allows the user to record the data for
      the individual tasks.

| **Record all form**                                                                                                          | **Record individual form**                                                                                                   |
|------------------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------|
| <img src="https://user-images.githubusercontent.com/4817531/222870875-75cb5f24-752c-4100-8883-1d2b48b71942.png" width="400"> | <img src="https://user-images.githubusercontent.com/4817531/222870896-3b95cff1-90bf-4529-b8cb-d2e816ed6b87.png" width="400"> |

- Sometimes immunizations have other immunizations as their prerequisites. Most of the time the dependent immunizations
  cannot be administered/recorded before the prerequisites are administered/recorded.
    - The dependent immunizations might have to satisfy a certain condition period after the prerequisites are given
      e.g. OPV1
      can only be administered after 28 days after OPV0
    - This means we need to dynamically change the dependent **DUE** date based on the prerequisites date of
      administration.

# How we intend to implement the above rules.

- Since the immunizations are recorded as Tasks we will be making the changes to satisfy the required rules on the
  tasking modules on FHIR Core.
- We will also define a new screen for upcoming encounters.
- The sets below are categorized as Code & Content Changes
    - **Code Changes** -- Code updates are needed to achieve the feature
    - **Content changes** -- Content changes are needed to achieve the feature

1. https://github.com/opensrp/fhircore/discussions/2141
2. https://github.com/opensrp/fhircore/discussions/2142
3. https://github.com/opensrp/fhircore/discussions/2140
4. https://github.com/opensrp/fhircore/discussions/2139
5. https://github.com/opensrp/fhircore/discussions/2143
6. https://github.com/opensrp/fhircore/discussions/2144
7. https://github.com/opensrp/fhircore/discussions/2765

# Assumptions

- The current profile configuration work is adequate to define the Upcoming Encounters screen.
- The immunizations are grouped based on the patient's age.
- The immunization **DUE**, **OVERDUE** and **EXPIRY** status are based on the patient's age.
- The StructureMap author has correctly written the extraction to provide all the information needed for the dependency
  mapping.

# Unscoped items

- Medical history page to show all the recorded/administered immunizations. @HenryRae and I to finalize its design then
  we can add it's scoping.

**Comment by pld** [comment]
Want to clarify some things, under code changes, when you say
- "The worker should only run if" isn't it that the worker always runs, but the below specifies what Tasks the worker searches for to update?
- Are these 2 bullets saying the same/an equivalent thing?
>    - The dependent task has an input.type attribute defined with
the {"coding":[{"system":"http://snomed.info/sct","code":"900000000000457003","display":"Reference set attribute (foundation metadata concept)"}]}.
>    - The Task should also have the prerequisite tasks referenced on the input attribute with the reference code
on the input.type attribute.
>        - Here is an example {"type":{"coding":[{"system":"http://snomed.info/sct","code":"900000000000457003","display":"Reference set attribute (foundation metadata concept)"}]},"reference":"Task/<string-uuid>"}.
- How is the "Immunization recording form" linked to the Task?
- Can we make explicit somewhere that the value in `Task.input.value` is in days?

For the "How the Record All form should work | Code or Content Changes" section, let's clarify that the task is to attempt to do with content only, if there's a blocker, report back to discuss, if we can't resolve we'll figure out a code solution, but not without sign-off.

For "Upcoming Encounters Screen | Code Changes" why can't we do this with the existing Profile configurations? Or, it sounds like we can w/this assumption, "The current profile configuration work is adequate to define the Upcoming Encounters screen.", what's the code change here?

**Comment by pld** [comment]
I'm gonna lock this, since you broke it down into multiple items
